README
This library is an implementation of the Secure Hash Algorithm 3
(SHA-3), also known as Keccak.  The implementation is constrained to
messages with an integral number of octets, i.e. sub-byte length
messages are not supported.

NOTE that prior to release 1.0.2 this package had a bug in the
generation of message digests where multiple calls to sha3-update
with partial buffers could lead to input data being ignored and
therefore erroneous message digests being generated.  Uses with
only one call to sha3-update and the high-level routines were not
affected by this bug.

NOTE that prior to release 1.1.0 this package computed digests
based on the Keccak submission to the SHA-3 contest and did not
yet take into account the added suffix that the FIPS 202 SHA-3
final standard adds to messages prior to calculating the digest,
since this was not part of the Keccak submission.  Starting with
1.1.0 the functions in the sha3 package do by default calculate
disgests that match the FIPS 202 standard, and will calculate
the old pre-standard digests only if the new optional keyword
argument :raw-keccak-p is passed with a true value.

The code should be portable across nearly all ANSI compliant CL
implementations with specialized versions tuned for implementations
that offer unboxed 64bit arithmetic, unboxed 32bit arithmetic and for
implementations with efficient fixnum arithmetic (requiring fixnums
that can represent (unsigned-byte 16)).  Especially the 64 and 32bit
implementations have been mostly optimized for SBCL and CMU CL.  For
those implementations, digests with a 1024 bit-rate (and 288 bit
digest output) can be generated in between 30 (64bit SBCL) to around
100 (32bit CMU CL) cycles/byte on an i7-640M; whereas optimized
C/assembler implementations reach around 12 to 50 cycles/byte on 64/32
bit Intel hardware.  The reason for the discrepancy probably lies in
missing peephole and dependency optimizations in the SBCL/CMU CL
compiler backend.

The mid-level interfaces to the digest routines are the functions

- sha3:sha3-init &key output-bit-length bit-rate
  
  Create and return a new SHA-3 state.  If `output-bit-length' is
  specified then the state will run at the bit rate specified for the
  given output bit length.  If `output-bit-length' is unspecified,
  `bit-rate' can be specified to select a suitable bit rate.  If both
  are left unspecified then a default bit rate of 1024 bits is
  selected, which is suitable for arbitrary output bit lengths of up
  to 288 bits.

- sha3:sha3-copy state
  
  Return an independent copy of the SHA-3 state `state'.

- sha3:sha3-state-p state
  
  Test whether a given object is a SHA-3 state, i.e. is an instance of
  the class sha3:sha3-state.

- sha3:sha3-update state vector &key (start 0) (end (length vector))
  
  Update the given SHA-3 state `state' from `vector', which must be a
  simple-array with element-type (unsigned-byte 8), bounded by `start'
  and `end', which must be numeric bounding-indices.

- sha3:sha3-final state &key output-bit-length raw-keccak-p
  
  If the given SHA-3 state `state' has not already been finalized,
  finalize it by processing any remaining input in its buffer, with
  the specified suffix of 01 and suitable padding as specified by the
  SHA-3 standard (the specified SHA-3 suffix can be elided with the
  optional keyword argument `raw-keccak-p' to generate digests as the
  initial Keccak submission would have generated).  Returns the
  message digest as a simple-array of (unsigned-byte 8).  The length
  of the returned digest is determined either by the output bit length
  or bit rate specified on state creation, or for the special case of
  default parameters being used, by the optional keyword argument
  `output-bit-length'.  If the state has previously been finalized,
  the function will return the digest again.

For convenience the following high-level functions produce digests in
one step from 1d simple-arrays and streams with element-type
(unsigned-byte 8), as well as files:

- sha3:sha3-digest-vector vector &key (start 0) end (output-bit-length 512) raw-keccak-p
  
  Calculate an SHA-3 message-digest of data in `vector', which should
  be a 1d simple-array with element type (unsigned-byte 8), bounded by
  `start' and `end'.  The bit length of the message digest produced is
  controlled by `output-bit-length', which can take on the values 224,
  256, 288, 384 and 512, which is the default value.  Using the optional
  `raw-keccak-p' keyword argument the SHA-3 mandated 01 suffix that is
  appended to the actual message prior to padding can be elided to yield
  message digests that match the original Keccak submission instead of
  the actual SHA-3 standard.  Use this option only for compatibility
  with historical implementations.

- sha3:sha3-digest-stream stream &key (output-bit-length 512) raw-keccak-p
  
  Calculate an SHA-3 message-digest of data read from `stream', which
  should be a stream with element type (unsigned-byte 8).  The bit
  length of the message digest produced is controlled by
  `output-bit-length', which can take on the values 224, 256, 288, 384
  and 512, which is the default value.  Using the optional `raw-keccak-p'
  keyword argument the SHA-3 mandated 01 suffix that is appended to the
  actual message prior to padding can be elided to yield message digests
  that match the original Keccak submission instead of the actual SHA-3
  standard.  Use this option only for compatibility with historical
  implementations.  

- sha3:sha3-digest-file pathname &key (output-bit-length 512) raw-keccak-p
  
  Calculate an SHA-3 message-digest of the file specified by
  `pathname'.  The bit length of the message digest produced is
  controlled by `output-bit-length', which can take on the values 224,
  256, 288, 384 and 512, which is the default value.  Using the optional
  `raw-keccak-p' keyword argument the SHA-3 mandated 01 suffix that is
  appended to the actual message prior to padding can be elided to yield
  message digests that match the original Keccak submission instead of
  the actual SHA-3 standard.  Use this option only for compatibility
  with historical implementations.

Note that in order to generate a message digest of a string it will
have to be converted to a simple-array with element-type
(unsigned-byte 8) in the proper output-encoding.  This will have to
rely on implementation-specific functions and is not part of the SHA3
library.

The file keccak-reference.lisp contains a slow simple reference
implementation, and testdriver code, which allows testing of the tuned
implementations against this reference and against test data available
from the Keccak Site at: http://keccak.noekeon.org/KeccakKAT-3.zip

The testcases from the Keccak test data can be run with the following
form:

(keccak:test-keccak-msgkat 
 "/Path/To/MsgKatDirectory"
 (lambda (total-bits bit-rate output-bits message)
   (declare (ignore total-bits bit-rate))
   (sha3:sha3-digest-vector message :output-bit-length output-bits :raw-keccak-p t)))

The adapted SHA-3 testcases from the Keccak Code Package test vectors
available under https://github.com/gvanas/KeccakCodePackage/tree/master/TestVectors
can be run with the following form:

(keccak:test-sha3-msgkat 
 "/Path/To/MsgKatDirectory"
 (lambda (total-bits bit-rate output-bits message)
   (declare (ignore total-bits bit-rate))
   (sha3:sha3-digest-vector message :output-bit-length output-bits)))

This SHA-3 implementation is licensed under the MIT-style license
contained in the file COPYING and the header of each source file.
Many thanks go to the Keccak Team (Guido Bertoni, Joan Daemen, Michaël
Peeters and Gilles Van Assche, cf. http://keccak.noekeon.org) for
their algorithm and excellent documentation and reference
implementations.

Please direct any feedback to pmai@pmsf.de.  A git repository of this
library is available under git://github.com/pmai/sha3.git


NEWS
Release 1.1.2
=============

 * Switch to 16bit implementation for 64bit LispWorks 8, to avoid
   compiler optimization problems in keccak-state-merge-input, and
   also slightly higher performance due to non-consing.

Release 1.1.1
=============

 * Fix a bug in pad-message-to-width on ACL, which would lead to
   completely erroneous results being calculated for all inputs
   due to differences in the result of adjust-array between ACL
   and all other currently supported implementations.

Release 1.1.0
=============

 * Change sha3 functions to match the final FIPS 202 SHA3 standard,
   by appending a 01 suffix to messages prior to digest calculation.
   The old Keccak submission behavior can be retained by passing in
   the new raw-keccak-p keyword argument with true to the relevant
   functions.  Thanks to David McClain for prompting this change.

 * Fixes a bug in the bug fix of release 1.0.2 for the sha3-update
   handling of not completely filled buffers reported by David McClain,
   which could lead to out-of-bounds accesses in calls to the
   pad-message-to-width function.

Release 1.0.2
=============

 * Fixes a bug reported by Orivej Desh where two or more calls to
   sha3-update which didn't fill the buffer could lead to the second
   and later updates being ignored, thereby creating wrong message
   digests.

Release 1.0.1
=============

 * Change to 32bit implementation for 64bit LispWorks.

Release 1.0.0
=============

 * Initial release with optimized versions for 64bit, 32bit and
   fixnum arithmetic.


common.lisp
;;;; SHA3 --- Secure Hash Algorithm 3 (Keccak) Implementation
;;;;
;;;; Copyright (C) 2012 -- 2018 PMSF IT Consulting Pierre R. Mai.
;;;;
;;;; Permission is hereby granted, free of charge, to any person obtaining
;;;; a copy of this software and associated documentation files (the
;;;; "Software"), to deal in the Software without restriction, including
;;;; without limitation the rights to use, copy, modify, merge, publish,
;;;; distribute, sublicense, and/or sell copies of the Software, and to
;;;; permit persons to whom the Software is furnished to do so, subject to
;;;; the following conditions:
;;;;
;;;; The above copyright notice and this permission notice shall be
;;;; included in all copies or substantial portions of the Software.
;;;;
;;;; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
;;;; EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
;;;; MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
;;;; IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY CLAIM, DAMAGES OR
;;;; OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
;;;; ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
;;;; OTHER DEALINGS IN THE SOFTWARE.
;;;;
;;;; Except as contained in this notice, the name of the author shall
;;;; not be used in advertising or otherwise to promote the sale, use or
;;;; other dealings in this Software without prior written authorization
;;;; from the author.
;;;; 
;;;; $Id$

(cl:in-package #:sha3)

;;;; %File Description:
;;;; 
;;;; This file contains common definitions and utility macros/functions
;;;; used in the specifically optimized implementations of keccak.
;;;; 

;;;
;;; Optimization
;;;

(eval-when (:compile-toplevel :load-toplevel :execute)
  (defvar *optimize-declaration* 
    '(optimize (speed 3) (space 0) (safety 0) (debug 0))
    "Global optimize declaration used for performance critical functions.
This can be changed prior to compiling the package for debugging/testing
purposes."))

;;;
;;; LOGXOR Reduction Hack for certain lisps
;;;

#+(or lispworks ccl)
(defun logxor (&rest args)
  (apply #'cl:logxor args))

#+(or lispworks ccl)
(define-compiler-macro logxor (&whole form &rest args)
  (labels ((binarify (list)
             (if (rest list)
                 `(cl:logxor ,(car list) ,(binarify (rest list)))
               (first list))))
    (if (null args)
        form
        (binarify args))))

;;;
;;; Partial Evaluation Helpers
;;;

(defun trivial-macroexpand-all (form env)
  "Trivial and very restricted code-walker used in partial evaluation macros.
Only supports atoms and function forms, no special forms."
  (let ((real-form (macroexpand form env)))
    (cond
      ((atom real-form)
       real-form)
      (t
       (list* (car real-form)
              (mapcar #'(lambda (x) (trivial-macroexpand-all x env)) 
                      (cdr real-form)))))))

(defmacro dotimes-unrolled ((var limit) &body body &environment env)
  "Unroll the loop body at compile-time."
  (loop for x from 0 below (eval (trivial-macroexpand-all limit env))
        collect
        `(symbol-macrolet ((,var ,x)) ,@body)
          into forms
        finally
     (return `(progn ,@forms))))

;;;
;;; Keccak-f-1600 definitions
;;;

(eval-when (:compile-toplevel :load-toplevel :execute)
  (defconstant +keccak-state-columns+ 5
    "Width of Keccak state in the x axis")
  (defconstant +keccak-state-rows+ 5
    "Width of Keccak state in the y axis")
  
  (defconstant +keccak-state-lanes+ (* +keccak-state-columns+ +keccak-state-rows+)
    "Total number of lanes in Keccak state")

  (defconstant +keccak-1600-lane-width+ 64
    "Lane width for Keccak-1600.")
  
  (defconstant +keccak-1600-lane-byte-width+ (truncate +keccak-1600-lane-width+ 8)
    "Lane width in bytes for Keccak-1600."))

(deftype keccak-1600-lane ()
  "Type of a keccak lane for Keccak-1600."
  `(unsigned-byte ,+keccak-1600-lane-width+))

;;;
;;; Keccak Constants
;;;

(defparameter *keccak-f-round-constants*
  (make-array '(24) :element-type 'keccak-1600-lane
              :initial-contents
              #-sha3-fixed-constants
              (loop with lfrstate = #x01
                    for i from 0 below 24
                    collect
                 (loop with const = 0
                       for j from 0 below 7
                       for bit-position = (1- (ash 1 j))
                       when (logbitp 0 lfrstate)
                         do (setq const (logxor const (ash 1 bit-position)))
                       do (setq lfrstate (if (logbitp 7 lfrstate)
                                             (logxor (ldb (byte 8 0) (ash lfrstate 1))
                                                     #x71)
                                             (ash lfrstate 1)))
                       finally (return const)))
              #+sha3-fixed-constants
              '(#x0000000000000001
                #x0000000000008082
                #x800000000000808a
                #x8000000080008000
                #x000000000000808b
                #x0000000080000001
                #x8000000080008081
                #x8000000000008009
                #x000000000000008a
                #x0000000000000088
                #x0000000080008009
                #x000000008000000a
                #x000000008000808b
                #x800000000000008b
                #x8000000000008089
                #x8000000000008003
                #x8000000000008002
                #x8000000000000080
                #x000000000000800a
                #x800000008000000a
                #x8000000080008081
                #x8000000000008080
                #x0000000080000001
                #x8000000080008008))
  "Keccak Round Constants")

(defparameter *keccak-f-rotate-offsets*
  (make-array (list +keccak-state-columns+ +keccak-state-rows+)
              :element-type '(unsigned-byte 8)
              :initial-contents
              '(( 0 36  3 41 18)
                ( 1 44 10 45  2)
                (62  6 43 15 61)
                (28 55 25 21 56)
                (27 20 39  8 14)))
  "Keccak Rotate Offsets")

(defmacro get-rotate-offset (x y &environment env)
  (aref *keccak-f-rotate-offsets* 
        (eval (trivial-macroexpand-all x env))
        (eval (trivial-macroexpand-all y env))))

;;;
;;; Message Padding for last block
;;;

(defun pad-message-to-width (message bit-width add-fips-202-suffix-p)
  "Destructively pad the given message to the given bit-width according to 
the Keccak 10*1 padding rules, optionally appending the FIPS 202/SHA-3
mandated 01 suffix first, and return the padded message."
  (let* ((message-byte-length (length message))
         (width-bytes (truncate bit-width 8))
         (padded (make-array width-bytes
                             :element-type '(unsigned-byte 8))))
    (replace padded message)
    ;; FIPS 202 SHA-3 mandates the appending of a 01 suffix prior to the
    ;; final Keccak padding so that the first byte following the message
    ;; will be #b00000101 instead of #b00000001 for raw Keccak.
    (setf (aref padded message-byte-length) (if add-fips-202-suffix-p #x06 #x01))
    (loop for index from (1+ message-byte-length) below width-bytes
          do (setf (aref padded index) #x00)
          finally
       (setf (aref padded (1- width-bytes))
             (logior #x80 (aref padded (1- width-bytes)))))
    padded))


pkgdef.lisp
;;;; SHA3 --- Secure Hash Algorithm 3 (Keccak) Implementation
;;;;
;;;; Copyright (C) 2012 -- 2018 PMSF IT Consulting Pierre R. Mai.
;;;;
;;;; Permission is hereby granted, free of charge, to any person obtaining
;;;; a copy of this software and associated documentation files (the
;;;; "Software"), to deal in the Software without restriction, including
;;;; without limitation the rights to use, copy, modify, merge, publish,
;;;; distribute, sublicense, and/or sell copies of the Software, and to
;;;; permit persons to whom the Software is furnished to do so, subject to
;;;; the following conditions:
;;;;
;;;; The above copyright notice and this permission notice shall be
;;;; included in all copies or substantial portions of the Software.
;;;;
;;;; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
;;;; EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
;;;; MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
;;;; IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY CLAIM, DAMAGES OR
;;;; OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
;;;; ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
;;;; OTHER DEALINGS IN THE SOFTWARE.
;;;;
;;;; Except as contained in this notice, the name of the author shall
;;;; not be used in advertising or otherwise to promote the sale, use or
;;;; other dealings in this Software without prior written authorization
;;;; from the author.
;;;; 
;;;; $Id$

;;;; %File Description:
;;;; 
;;;; This file contains the package definition
;;;;

(cl:defpackage #:sha3
  (:use #:common-lisp)
  #+(or lispworks ccl)
  (:shadow #:logxor)
  (:export 
   ;; Mid-Level
   #:sha3-state #:sha3-state-p
   #:sha3-init #:sha3-copy #:sha3-update #:sha3-final
   ;; High-Level
   #:sha3-digest-vector #:sha3-digest-stream #:sha3-digest-file
   ))


sha3.lisp
;;;; SHA3 --- Secure Hash Algorithm 3 (Keccak) Implementation
;;;;
;;;; Copyright (C) 2012 -- 2018 PMSF IT Consulting Pierre R. Mai.
;;;;
;;;; Permission is hereby granted, free of charge, to any person obtaining
;;;; a copy of this software and associated documentation files (the
;;;; "Software"), to deal in the Software without restriction, including
;;;; without limitation the rights to use, copy, modify, merge, publish,
;;;; distribute, sublicense, and/or sell copies of the Software, and to
;;;; permit persons to whom the Software is furnished to do so, subject to
;;;; the following conditions:
;;;;
;;;; The above copyright notice and this permission notice shall be
;;;; included in all copies or substantial portions of the Software.
;;;;
;;;; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
;;;; EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
;;;; MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
;;;; IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY CLAIM, DAMAGES OR
;;;; OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
;;;; ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
;;;; OTHER DEALINGS IN THE SOFTWARE.
;;;;
;;;; Except as contained in this notice, the name of the author shall
;;;; not be used in advertising or otherwise to promote the sale, use or
;;;; other dealings in this Software without prior written authorization
;;;; from the author.
;;;; 
;;;; $Id$

(cl:in-package #:sha3)

;;;; %File Description:
;;;; 
;;;; This file contains the implementation of mid- and high-level
;;;; SHA-3 functions based on the optimized Keccak implementations.
;;;; 
;;;; The main entry points are the mid-level functions sha3-init,
;;;; sha3-update and sha3-final to initialize, update and finalize an
;;;; sha3 hash, and sha3-copy in order to copy a sha3 state.
;;;;
;;;; For convenience high-level functions to hash a given vector,
;;;; stream or file are provided as sha3-digest-vector,
;;;; sha3-digest-stream and sha3-digest-vector.
;;;;

;;;
;;; Mid-Level Routines
;;;

(defstruct (sha3-state
             (:constructor make-sha3-state (bit-rate))
             (:copier))
  (state (make-keccak-1600-state) :type keccak-1600-state :read-only t)
  (bit-rate 1024 :type (integer 0 1600) :read-only t)
  (buffer (make-array 200 :element-type '(unsigned-byte 8)) :read-only t
         :type (simple-array (unsigned-byte 8) (200)))
  (buffer-index 0 :type (integer 0 199))
  (finalized-p nil))

(defun sha3-init (&key (output-bit-length nil output-bit-length-p)
                  (bit-rate (if (and output-bit-length-p
                                     output-bit-length)
                                (- 1600 (* 2 output-bit-length))
                                1024)))
  "Create and return a new SHA-3 state.  If `output-bit-length' is specified
then the state will run at the bit rate specified for the given output bit
length.  If `output-bit-length' is unspecified, `bit-rate' can be specified
to select a suitable bit rate.  If both are left unspecified then a default
bit rate of 1024 bits is selected, which is suitable for arbitrary output
bit lengths of up to 288 bits."
  (check-type output-bit-length (member nil 224 256 288 384 512)
              "Legal output-bit-length specifier for SHA-3/Keccak-1600")
  (check-type bit-rate (member 576 832 1024 1088 1152)
              "Legal bit-rate for SHA-3/Keccak-1600")
  (if (or (null output-bit-length)
          (= bit-rate (- 1600 (* 2 output-bit-length)))
          (= bit-rate 1024))
      (make-sha3-state bit-rate)
      (error "Illegal combination of output-bit-length ~D and bit-rate ~D."
             output-bit-length bit-rate)))

(defun sha3-copy (state)
  "Return an independent copy of the SHA-3 state `state'."
  (let ((result (make-sha3-state (sha3-state-bit-rate state))))
    (setf (sha3-state-buffer-index result) (sha3-state-buffer-index state)
          (sha3-state-finalized-p result) (sha3-state-finalized-p state))
    (replace (sha3-state-buffer result) (sha3-state-buffer state))
    (replace (sha3-state-state result) (sha3-state-state state))
    result))

(defun sha3-update (state vector &key (start 0) (end (length vector)))
  "Update the given SHA-3 state `state' from `vector', which must be
a simple-array with element-type (unsigned-byte 8), bounded by `start'
and `end', which must be numeric bounding-indices."
  (declare (type sha3-state state) 
           (type (simple-array (unsigned-byte 8) (*)) vector)
           (type fixnum start end)
           (optimize (speed 3) (safety 1) (space 0) (debug 1)))
  (let* ((keccak-state (sha3-state-state state))
         (buffer (sha3-state-buffer state))
         (buffer-index (sha3-state-buffer-index state))
         (bit-rate (sha3-state-bit-rate state))
         (rate-bytes (truncate bit-rate 8)))
    (declare (type keccak-1600-state keccak-state)
             (type (simple-array (unsigned-byte 8) (200)) buffer)
             (type (integer 0 199) buffer-index)
             (type (integer 0 1600) bit-rate)
             (type (integer 0 200) rate-bytes)
             #.*optimize-declaration*)
    ;; Handle potential remaining bytes
    (unless (zerop buffer-index)
      (let ((remainder (- rate-bytes buffer-index))
            (length (- end start)))
        (declare (type fixnum remainder length))
        (replace buffer vector :start1 buffer-index :start2 start :end2 end)
        ;; Return if still unfilled buffer
        (when (< length remainder)
          (incf (sha3-state-buffer-index state) length)
          (return-from sha3-update))
        ;; Else handle now complete buffer
        (keccak-state-merge-input keccak-state bit-rate buffer 0)
        (keccak-f keccak-state)
        (setf (sha3-state-buffer-index state) 0
              start (+ start remainder))))
    ;; Now handle full blocks, stuff any remainder into buffer
    (loop for block-offset of-type fixnum from start below end by rate-bytes
          do
       (cond
         ((<= (+ block-offset rate-bytes) end)
          (keccak-state-merge-input keccak-state bit-rate vector block-offset)
          (keccak-f keccak-state))
         (t
          (replace buffer vector :start1 0 :start2 block-offset)
          (setf (sha3-state-buffer-index state) (- end block-offset)))))))

(defun sha3-final (state &key (output-bit-length nil output-bit-length-p) raw-keccak-p)
  "If the given SHA-3 state `state' has not already been finalized,
finalize it by processing any remaining input in its buffer, with
the specified suffix of 01 and suitable padding as specified by the
SHA-3 standard (the specified SHA-3 suffix can be elided with the
optional keyword argument `raw-keccak-p' to generate digests as the
initial Keccak submission would have generated).   Returns the
message digest as a simple-array of (unsigned-byte 8).  The length
of the returned digest is determined either by the output bit length
or bit rate specified on state creation, or for the special case of
default parameters being used, by the optional keyword argument
`output-bit-length'.  If the state has previously been finalized,
the function will return the digest again."
  (declare (type sha3-state state)
           (type (or null (integer 0 1600)) output-bit-length)
           (optimize (speed 3) (safety 1) (space 0) (debug 1)))
  (let ((keccak-state (sha3-state-state state))
        (buffer (sha3-state-buffer state))
        (buffer-index (sha3-state-buffer-index state))
        (bit-rate (sha3-state-bit-rate state))
        (finalized-p (sha3-state-finalized-p state)))
    (declare (type keccak-1600-state keccak-state)
             (type (simple-array (unsigned-byte 8) (200)) buffer)
             (type (integer 0 199) buffer-index)
             (type (integer 0 1600) bit-rate)
             (type (or null (simple-array (unsigned-byte 8) (*))) finalized-p)
             #.*optimize-declaration*)
    ;; Determine output-bit-length
    (if output-bit-length-p
        (unless (or (= bit-rate 1024)
                    (= (* 2 output-bit-length) (- 1600 bit-rate)))
          (error "Illegal output-bit-length ~D specified!" output-bit-length))
        (setq output-bit-length (truncate (- 1600 bit-rate) 2)))
    (cond
      ;; Check if already finalized
      (finalized-p
       (unless (= (* (length finalized-p) 8) output-bit-length)
         (error "Mismatch in output-bit-length ~D in repeated call to sha3-final! ~
                 Should be: ~D!"
                output-bit-length (* (length finalized-p) 8)))
       finalized-p)
      ;; Finalize
      (t
       (keccak-state-merge-input keccak-state bit-rate 
                                 (pad-message-to-width
                                  (subseq buffer 0 buffer-index)
                                  bit-rate
                                  (not raw-keccak-p))
                                 0)
       (keccak-f keccak-state)
       (setf (sha3-state-buffer-index state) 0
             (sha3-state-finalized-p state)
             (keccak-state-extract-output keccak-state output-bit-length))))))

;;;
;;; High-Level Routines
;;;

(defun sha3-digest-vector (vector &key (start 0) end 
                           (output-bit-length 512)
                           raw-keccak-p)
  "Calculate an SHA-3 message-digest of data in `vector', which should
be a 1d simple-array with element type (unsigned-byte 8), bounded by
`start' and `end'.  The bit length of the message digest produced is
controlled by `output-bit-length', which can take on the values 224,
256, 288, 384 and 512, which is the default value.  Using the optional
`raw-keccak-p' keyword argument the SHA-3 mandated 01 suffix that is
appended to the actual message prior to padding can be elided to yield
message digests that match the original Keccak submission instead of
the actual SHA-3 standard.  Use this option only for compatibility
with historical implementations."
  (declare (optimize (speed 3) (safety 3) (space 0) (debug 1))
           (type (simple-array (unsigned-byte 8) (*)) vector)
           (type fixnum start)
           (type (or null fixnum) end)
           (type (integer 0 1600) output-bit-length))
  (locally
      (declare (optimize (safety 1) (debug 0)))
    (let ((state (sha3-init :output-bit-length output-bit-length)))
      (declare (type sha3-state state))
      (let ((real-end (or end (length vector))))
        (declare (type fixnum real-end))
        (sha3-update state vector :start start :end real-end))
      (sha3-final state :raw-keccak-p raw-keccak-p))))

(eval-when (:compile-toplevel :load-toplevel :execute)
  (defconstant +buffer-size+ (* 128 1024)
    "Size of internal buffer to use for `sha3-digest-stream' and 
`sha3-digest-file' operations."))

(deftype buffer-index () `(integer 0 ,+buffer-size+))

(defun sha3-digest-stream (stream &key (output-bit-length 512) raw-keccak-p)
  "Calculate an SHA-3 message-digest of data read from `stream', which
should be a stream with element type (unsigned-byte 8).  The bit
length of the message digest produced is controlled by
`output-bit-length', which can take on the values 224, 256, 288, 384
and 512, which is the default value.  Using the optional `raw-keccak-p'
keyword argument the SHA-3 mandated 01 suffix that is appended to the
actual message prior to padding can be elided to yield message digests
that match the original Keccak submission instead of the actual SHA-3
standard.  Use this option only for compatibility with historical
implementations."
  (declare (optimize (speed 3) (safety 3) (space 0) (debug 1))
           (type stream stream)
           (type (integer 0 1600) output-bit-length))
  (locally
      (declare (optimize (safety 1) (debug 0)))
    (unless (equal (stream-element-type stream) '(unsigned-byte 8))
      (error "Illegal stream-element-type ~S, must be ~S."
             (stream-element-type stream) '(unsigned-byte 8)))
    (let ((buffer (make-array '(#.+buffer-size+) :element-type '(unsigned-byte 8)))
          (state (sha3-init :output-bit-length output-bit-length)))
      (declare (type sha3-state state)
               (type (simple-array (unsigned-byte 8) (#.+buffer-size+)) buffer))
      (loop for bytes of-type buffer-index = (read-sequence buffer stream)
            do (sha3-update state buffer :end bytes)
            until (< bytes +buffer-size+)
            finally
         (return (sha3-final state :raw-keccak-p raw-keccak-p))))))

(defun sha3-digest-file (pathname &key (output-bit-length 512) raw-keccak-p)
  "Calculate an SHA-3 message-digest of the file specified by
`pathname'.  The bit length of the message digest produced is
controlled by `output-bit-length', which can take on the values 224,
256, 288, 384 and 512, which is the default value.  Using the optional
`raw-keccak-p' keyword argument the SHA-3 mandated 01 suffix that is
appended to the actual message prior to padding can be elided to yield
message digests that match the original Keccak submission instead of
the actual SHA-3 standard.  Use this option only for compatibility
with historical implementations."
  (declare (optimize (speed 3) (safety 3) (space 0) (debug 1))
           (type (integer 0 1600) output-bit-length))
  (locally
      (declare (optimize (safety 1) (debug 0)))
    (with-open-file (stream pathname :element-type '(unsigned-byte 8))
      (sha3-digest-stream stream :output-bit-length output-bit-length :raw-keccak-p raw-keccak-p))))
